peps/pep-0728.rst

```diff
@@ -141,424 +142,422 @@ in TypeScript achieve this:
 
-    Movie = TypedDict("Movie", {"name": str, "__extra_items__": bool}, closed=True)
+    Movie = TypedDict("Movie", {"name": str}, extra_items=bool)
 
-Accessing extra keys is allowed. Type checkers must infer its value type from
-the value type of ``__extra_items__``::
+Accessing extra items is allowed. Type checkers must infer their value type from
+the ``extra_items`` argument::
 
     def f(movie: Movie) -> None:
         reveal_type(movie["name"])              # Revealed type is 'str'
         reveal_type(movie["novel_adaptation"])  # Revealed type is 'bool'
-
-When a TypedDict type defines ``__extra_items__`` without ``closed=True``,
-``closed`` defaults to ``False`` and the key is assumed to be a regular key::
-
-    class Movie(TypedDict):
-        name: str
-        __extra_items__: bool
     
-    a: Movie = {"name": "Blade Runner", "novel_adaptation": True}  # Not OK. Unexpected key 'novel_adaptation'
-    b: Movie = {
-        "name": "Blade Runner",
-        "__extra_items__": True,  # OK
-    }
-
-For such non-closed TypedDict types, it is assumed that they allow non-required
-extra items of value type ``ReadOnly[object]`` during inheritance or type
-consistency checks. However, extra keys found during construction should still
-be rejected by the type checker.
+``extra_items`` is inherited through subclassing::
 
-``closed`` is not inherited through subclassing::
-
-    class MovieBase(TypedDict, closed=True):
+    class MovieBase(TypedDict, extra_items=int | None):
         name: str
-        __extra_items__: ReadOnly[str | None]
     
     class Movie(MovieBase):
-        __extra_items__: str  # A regular key
-    
-    a: Movie = {"name": "Blade Runner", "__extra_items__": None}  # Not OK. 'None' is incompatible with 'str'
+        year: int
+
+    a: Movie = {"name": "Blade Runner", "year": None}  # Not OK. 'None' is incompatible with 'int'
     b: Movie = {
         "name": "Blade Runner",
-        "__extra_items__": "A required regular key",
+        "year": 1982,
         "other_extra_key": None,
     }  # OK
 
-Here, ``"__extra_items__"`` in ``a`` is a regular key defined on ``Movie`` where
-its value type is narrowed from ``ReadOnly[str | None]`` to ``str``,
-``"other_extra_key"`` in ``b`` is an extra key whose value type must be
-consistent with the value type of ``"__extra_items__"`` defined on
-``MovieBase``.
+Here, ``'year'`` in ``a`` is an extra key defined on ``Movie`` whose value type
+is ``int``. ``'other_extra_key'`` in ``b`` is another extra key whose value type
+must be assignable to the value of ``extra_items`` defined on ``MovieBase``.
+
+The ``closed`` Class Parameter
+------------------------------
+
+When ``closed=True`` is set, no extra items are allowed. This is a shorthand for
+``extra_items=Never``, because there can't be a value type that is assignable to
+:class:`~typing.Never`.
+
+Different from ``total``, only a literal ``True`` is supported as the value of
```

Comment:
Feel we should allow `closed=False`, just if people want to be explicit. However, the value should be a literal.

Reply:
Thanks for noticing that. If `closed=False` means that the TypedDict is non-closed, we probably still need to default to unset, because `closed=True` would be inherited. We can make `closed=False` equivalent to unset, but it might not be intuitive.

It is kind of tricky to define the semantics of `closed=False` without making it confusing, but there might be a better way around this. This might be a good topic to discuss in public for the next revision.

Reply:
I'm not sure I see the problem; `closed=False` should be the default, so passing it explicitly should match the default behavior. A `closed=True` TypedDict could subclass a `closed=False` one, just as it can subclass a standard TypedDict without special options.

Reply:
I see. In this case, a `closed=False` TypedDict should be allowed to subclass a `closed=True` TypedDict, but such a TypedDict will remain closed as it inherits `extra_items` from the supertype. Could that be a source of confusion?